In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Establecer la opción para mostrar todas las columnas
pd.set_option('display.max_columns', None)

Cargamos el primer archivo con el que vamos a trabajar.

In [2]:
movies_raw_csv = pd.read_csv('data/movies_dataset.csv',low_memory=False)
movies_raw_csv.info()
movies_raw_csv.head(2)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45466 non-null  object 
 1   belongs_to_collection  4494 non-null   object 
 2   budget                 45466 non-null  object 
 3   genres                 45466 non-null  object 
 4   homepage               7782 non-null   object 
 5   id                     45466 non-null  object 
 6   imdb_id                45449 non-null  object 
 7   original_language      45455 non-null  object 
 8   original_title         45466 non-null  object 
 9   overview               44512 non-null  object 
 10  popularity             45461 non-null  object 
 11  poster_path            45080 non-null  object 
 12  production_companies   45463 non-null  object 
 13  production_countries   45463 non-null  object 
 14  release_date           45379 non-null  object 
 15  re

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   

                               homepage    id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story   862  tt0114709                en   
1                                   NaN  8844  tt0113497                en   

  original_title                                           overview  \
0      Toy Story  Led by Woody, Andy's toys live happily in his ...   
1        Jumanji  When siblings Judy and Peter discover an encha...   

  popularity                       poster_path  \
0  21.946943  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  17.015539  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   

                                production_countries release_date  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-10-30   
1  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-15   

       revenue  runtime                                   spoken_languages  \
0  373554033.0     81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0    104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   

     status                                    tagline      title  video  \
0  Released                                        NaN  Toy Story  False   
1  Released  Roll the dice and unleash the excitement!    Jumanji  False   

   vote_average  vote_count  
0           7.7      5415.0  
1           6.9      2413.0

In [3]:
(
    movies_raw_csv
    .isnull()
    .sum()
)

adult                        0
belongs_to_collection    40972
budget                       0
genres                       0
homepage                 37684
id                           0
imdb_id                     17
original_language           11
original_title               0
overview                   954
popularity                   5
poster_path                386
production_companies         3
production_countries         3
release_date                87
revenue                      6
runtime                    263
spoken_languages             6
status                      87
tagline                  25054
title                        6
video                        6
vote_average                 6
vote_count                   6
dtype: int64

Tenemos una base de datos de peliculas, la cual contiene 45466 filas y 24 columnas.
Las columnas 'belongs_to_collection', 'genres', 'production_companies','production_countries' y 'spoken_languages' poseen datos anidados.
Analizando la información del df podemos notar que 6 peliculas tienen valores nulos en el campo 'title' pero no en el campo 'original_title'. Buscaremos esas filas.

Valores nulos

In [4]:
movies_raw_csv[movies_raw_csv['title'].isnull()]

adult  \
19729                                              False   
19730                                 - Written by Ørnås   
29502                                              False   
29503   Rune Balot goes to a casino connected to the ...   
35586                                              False   
35587   Avalanche Sharks tells the story of a bikini ...   

                                   belongs_to_collection  \
19729                                                NaN   
19730                                           0.065736   
29502  {'id': 122661, 'name': 'Mardock Scramble Colle...   
29503                                           1.931659   
35586                                                NaN   
35587                                           2.185485   

                                 budget  \
19729                                 0   
19730  /ff9qCepilowshEtG2GYWwzt2bs4.jpg   
29502                                 0   
29503  /zV8bHuSL6WXoD6FWogP9j4x80bL.jpg   
35586                                 0   
35587  /zaSf5OG7V8X8gqFvly88zDdRm46.jpg   

                                                  genres  \
19729  [{'id': 28, 'name': 'Action'}, {'id': 53, 'nam...   
19730  [{'name': 'Carousel Productions', 'id': 11176}...   
29502  [{'id': 16, 'name': 'Animation'}, {'id': 878, ...   
29503  [{'name': 'Aniplex', 'id': 2883}, {'name': 'Go...   
35586  [{'id': 10770, 'name': 'TV Movie'}, {'id': 28,...   
35587  [{'name': 'Odyssey Media', 'id': 17161}, {'nam...   

                                                homepage          id  \
19729                                                NaN       82663   
19730  [{'iso_3166_1': 'CA', 'name': 'Canada'}, {'iso...  1997-08-20   
29502                      http://m-scramble.jp/exhaust/      122662   
29503  [{'iso_3166_1': 'US', 'name': 'United States o...  2012-09-29   
35586                                                NaN      249260   
35587           [{'iso_3166_1': 'CA', 'name': 'Canada'}]  2014-01-01   

         imdb_id original_language                            original_title  \
19729  tt0113002                en                              Midnight Man   
19730          0             104.0  [{'iso_639_1': 'en', 'name': 'English'}]   
29502  tt2423504                ja                          マルドゥック・スクランブル 排気   
29503          0              68.0      [{'iso_639_1': 'ja', 'name': '日本語'}]   
35586  tt2622826                en                          Avalanche Sharks   
35587          0              82.0  [{'iso_639_1': 'en', 'name': 'English'}]   

                                                overview  \
19729  British soldiers force a recently captured IRA...   
19730                                           Released   
29502         Third film of the Mardock Scramble series.   
29503                                           Released   
35586  A group of skiers are terrorized during spring...   
35587                                           Released   

                  popularity                          poster_path  \
19729                    NaN                                  NaN   
19730                    NaN                         Midnight Man   
29502                    NaN                                  NaN   
29503                    NaN  Mardock Scramble: The Third Exhaust   
35586                    NaN                                  NaN   
35587  Beware Of Frost Bites                     Avalanche Sharks   

      production_companies production_countries release_date  revenue  \
19729                  NaN                  NaN          NaN      NaN   
19730                False                  6.0            1      NaN   
29502                  NaN                  NaN          NaN      NaN   
29503                False                  7.0           12      NaN   
35586                  NaN                  NaN          NaN      NaN   
35587                False                  4.3           22      NaN   

   

Estas 6 filas también se eliminaran ya que la mayoría de sus datos son nulos o inconsistentes.

In [5]:
indice = movies_raw_csv[movies_raw_csv['title'].isnull()].index
indice

Int64Index([19729, 19730, 29502, 29503, 35586, 35587], dtype='int64')

In [6]:
movies = movies_raw_csv.drop(index=indice, inplace=False)

In [7]:
movies_raw_csv.shape

(45466, 24)

In [8]:
#Comprobamos que se eliminaron las 6 filas.
movies.shape

(45460, 24)

Valores duplicados

In [9]:
#filas duplicadas
(
    movies_raw_csv
    .duplicated(keep=False)
    .sum()
)

33

In [10]:
# Filas con campo id duplicado /  Una misma pelicula aparece varias veces en el dataframe.
(
    movies_raw_csv
    .duplicated(subset='id',keep=False)
    .sum()
)

59

In [11]:
(
    movies
    .duplicated(subset='id',keep=False)
    .sum()
)

59

In [12]:
movies_raw_csv.shape

(45466, 24)

In [13]:
movies = movies.drop_duplicates()

movies = movies.drop_duplicates(subset='id',keep=False)

In [14]:
movies.shape

(45417, 24)

In [15]:
#cambiamos los tipos de datos de las columnas
movies = movies.rename({'id': 'id_movie'}, axis=1)

In [16]:
movies['title'] = movies['title'].apply(lambda x: x.title())

In [17]:
movies = movies.astype({'id_movie' : 'int64', 
                        # 'title':'str',
                        # 'tagline':'str',
                        # 'overview':'str',
                        'original_language':'category',
                        # # 'runtime':'int', # posee valores nan por lo que aun no podemos cambiarlo a int
                        'status': 'category', 
                        # 'release_date': ,  # se tratará mas adelante
                        'budget': 'int64', 
                        'revenue': 'int64', 
                        'vote_count': 'int64', 
                        'vote_average': 'int', 
                        # 'popularity'
                        }, copy=False, errors='raise')

------------------

# Requerimientos

1- (x) Eliminar las columnas que no serán utilizadas, video,imdb_id,adult,original_title,poster_path y homepage. 

2- (X) Los valores nulos de los campos revenue, budget deben ser rellenados por el número 0.

3- (x) Los valores nulos del campo release date deben eliminarse.

4- (x) Crear la columna con el retorno de inversión, llamada return con los campos revenue y budget, dividiendo estas dos últimas revenue / budget, cuando no hay datos disponibles para calcularlo, deberá tomar el valor 0.

5- (x) De haber fechas, deberán tener el formato AAAA-mm-dd, además deberán crear la columna release_year donde extraerán el año de la fecha de estreno.

6- Algunos campos, como belongs_to_collection, production_companies y otros (ver diccionario de datos) están anidados, esto es o bien tienen un diccionario o una lista como valores en cada fila, ¡deberán desanidarlos y unirlos al dataset de nuevo para hacer alguna de las consultas de la API! O bien buscar la manera de acceder a esos datos sin desanidarlos.. (ver 3_creacion_de_tablas.ipynb)

----------------

1- Eliminar las columnas video, imdb_id,adult, original_title, poster_path y homepage. 

In [18]:
movies.columns

Index(['adult', 'belongs_to_collection', 'budget', 'genres', 'homepage',
       'id_movie', 'imdb_id', 'original_language', 'original_title',
       'overview', 'popularity', 'poster_path', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'video',
       'vote_average', 'vote_count'],
      dtype='object')

In [19]:
movies = movies[['id_movie','title','tagline','overview','genres','runtime', 'release_date', 'status',
                 'belongs_to_collection',
                 'original_language', 'spoken_languages',
                 'production_companies', 'production_countries',
                 'popularity','vote_average', 'vote_count',
                 'budget','revenue'
]]

In [20]:
movies.head(3)

id_movie             title  \
0       862         Toy Story   
1      8844           Jumanji   
2     15602  Grumpier Old Men   

                                             tagline  \
0                                                NaN   
1          Roll the dice and unleash the excitement!   
2  Still Yelling. Still Fighting. Still Ready for...   

                                            overview  \
0  Led by Woody, Andy's toys live happily in his ...   
1  When siblings Judy and Peter discover an encha...   
2  A family wedding reignites the ancient feud be...   

                                              genres  runtime release_date  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...     81.0   1995-10-30   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...    104.0   1995-12-15   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...    101.0   1995-12-22   

     status                              belongs_to_collection  \
0  Released  {'id': 10194, 'name': 'Toy Story Collection', ...   
1  Released                                                NaN   
2  Released  {'id': 119050, 'name': 'Grumpy Old Men Collect...   

  original_language                                   spoken_languages  \
0                en           [{'iso_639_1': 'en', 'name': 'English'}]   
1                en  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2                en           [{'iso_639_1': 'en', 'name': 'English'}]   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   

                                production_countries popularity  vote_average  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...  21.946943             7   
1  [{'iso_3166_1': 'US', 'name': 'United States o...  17.015539             6   
2  [{'iso_3166_1': 'US', 'name': 'United States o...    11.7129             6   

   vote_count    budget    revenue  
0        5415  30000000  373554033  
1        2413  65000000  262797249  
2          92         0          0

Ahora tenemos un df de 45407 filas y 18 columnas.

-------------------

2- Los valores nulos de los campos revenue, budget deben ser rellenados por el número 0.

In [21]:
movies['budget'].fillna(0, inplace=True)
movies['revenue'].fillna(0, inplace=True)

---

3- Los valores nulos del campo release date deben eliminarse.

In [22]:
movies = movies.dropna(subset='release_date')

In [23]:
movies.tail(4)

id_movie                title                 tagline  \
45462    111109  Century Of Birthing                     NaN   
45463     67758             Betrayal  A deadly game of wits.   
45464    227506     Satan Triumphant                     NaN   
45465    461257             Queerama                     NaN   

                                                overview  \
45462  An artist struggles to finish his work while a...   
45463  When one of her hits goes wrong, a professiona...   
45464  In a small town live two brothers, one a minis...   
45465  50 years after decriminalisation of homosexual...   

                                                  genres  runtime  \
45462                      [{'id': 18, 'name': 'Drama'}]    360.0   
45463  [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...     90.0   
45464                                                 []     87.0   
45465                                                 []     75.0   

      release_date    status belongs_to_collection original_language  \
45462   2011-11-17  Released                   NaN                tl   
45463   2003-08-01  Released                   NaN                en   
45464   1917-10-21  Released                   NaN                en   
45465   2017-06-09  Released                   NaN                en   

                               spoken_languages  \
45462         [{'iso_639_1': 'tl', 'name': ''}]   
45463  [{'iso_639_1': 'en', 'name': 'English'}]   
45464                                        []   
45465  [{'iso_639_1': 'en', 'name': 'English'}]   

                                    production_companies  \
45462             [{'name': 'Sine Olivia', 'id': 19653}]   
45463  [{'name': 'American World Pictures', 'id': 6165}]   
45464               [{'name': 'Yermoliev', 'id': 88753}]   
45465                                                 []   

                                    production_countries popularity  \
45462      [{'iso_3166_1': 'PH', 'name': 'Philippines'}]   0.178241   
45463  [{'iso_3166_1': 'US', 'name': 'United States o...   0.903007   
45464           [{'iso_3166_1': 'RU', 'name': 'Russia'}]   0.003503   
45465   [{'iso_3166_1': 'GB', 'name': 'United Kingdom'}]   0.163015   

       vote_average  vote_count  budget  revenue  
45462             9           3       0        0  
45463             3           6       0        0  
45464             0           0       0        0  
45465             0           0       0        0

---------------

4- Crear la columna con el retorno de inversión, llamada return con los campos revenue y budget, dividiendo estas dos últimas revenue / budget, cuando no hay datos disponibles para calcularlo, deberá tomar el valor 0.

In [24]:
# funcion que permite calcular el return
def ret(df,col1,col2):
    if df[col1] == 0:
        out = 0

    else:    
        out = df[col2] / df[col1]
    return out


In [25]:
# creamos la columna return
movies.insert(movies.columns.get_loc('revenue') + 1, 'return', movies.apply(ret,axis=1,args={'budget','revenue'}))

In [26]:
movies.head(2)

id_movie      title                                    tagline  \
0       862  Toy Story                                        NaN   
1      8844    Jumanji  Roll the dice and unleash the excitement!   

                                            overview  \
0  Led by Woody, Andy's toys live happily in his ...   
1  When siblings Judy and Peter discover an encha...   

                                              genres  runtime release_date  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...     81.0   1995-10-30   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...    104.0   1995-12-15   

     status                              belongs_to_collection  \
0  Released  {'id': 10194, 'name': 'Toy Story Collection', ...   
1  Released                                                NaN   

  original_language                                   spoken_languages  \
0                en           [{'iso_639_1': 'en', 'name': 'English'}]   
1                en  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   

                                production_countries popularity  vote_average  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...  21.946943             7   
1  [{'iso_3166_1': 'US', 'name': 'United States o...  17.015539             6   

   vote_count    budget    revenue    return  
0        5415  30000000  373554033  0.080310  
1        2413  65000000  262797249  0.247339

------------

5- De haber fechas, deberán tener el formato AAAA-mm-dd, además deberán crear la columna release_year donde extraerán el año de la fecha de estreno.

In [27]:
#cambiamos los tipos de datos de la columna 'release_date'
movies['release_date'] = pd.to_datetime(movies['release_date'], format='%Y-%m-%d',yearfirst=True, errors='coerce')

In [28]:
#Creamos la columna 'release_year'
movies.insert(movies.columns.get_loc('release_date') + 1, 'release_year', movies['release_date'].dt.year)

In [29]:
movies.head(2)

id_movie      title                                    tagline  \
0       862  Toy Story                                        NaN   
1      8844    Jumanji  Roll the dice and unleash the excitement!   

                                            overview  \
0  Led by Woody, Andy's toys live happily in his ...   
1  When siblings Judy and Peter discover an encha...   

                                              genres  runtime release_date  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...     81.0   1995-10-30   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...    104.0   1995-12-15   

   release_year    status                              belongs_to_collection  \
0          1995  Released  {'id': 10194, 'name': 'Toy Story Collection', ...   
1          1995  Released                                                NaN   

  original_language                                   spoken_languages  \
0                en           [{'iso_639_1': 'en', 'name': 'English'}]   
1                en  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   

                                production_countries popularity  vote_average  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...  21.946943             7   
1  [{'iso_3166_1': 'US', 'name': 'United States o...  17.015539             6   

   vote_count    budget    revenue    return  
0        5415  30000000  373554033  0.080310  
1        2413  65000000  262797249  0.247339

------------

In [30]:
#crear el csv 
movies.to_csv("data/movies_2.csv",index=False)